### ISSUE TITLE
BigQuery lets you access other datasets when writing SQL queries

### ISSUE THREAD

**Issue by VikramTiwari** [metabase/metabase#5274]
Basics:
- Your databases: BigQuery
- Metabase version: 0.24.1
- Metabase internal database: MySQL

Steps to reproduce:
- 2 users in 2 separate groups with 2 different datasets from same BigQuery database
- using same backend credentials to connect to BigQuery database
- using SQL query builder
- user from group A can query dataset/tables from group B even though Metabase permissions restrict them to do so
- it works cause the backend credentials to BigQuery are for a project/database level which gives access to all the datasets in that DB, even though those datasets might be for different user groups on Metabase
- Metabase should restrict query transfer to BigQuery if the group does not have access to query that dataset

**Comment by camsaul** [comment]
@VikramTiwari since BigQuery lets you access other datasets with SQL you might want to turn off SQL access for any users who you don't want to be able to do that

**Comment by VikramTiwari** [comment]
I think the current structure makes sense. Since most of the times, we want to aggregate data from all the teams into one project (DB) where each team might have access to different datasets.

To overcome this scenario an easy fix would be to check for the dataset before passing that forward to BigQuery.

**Comment by camsaul** [comment]
that would require us to parse the SQL to see which dataset is referenced which is probably too fragile for security purposes

**Comment by camsaul** [comment]
i think it makes more sense to treat different datasets as different schemas within one database. You could still restrict access to various datasets in the GUI Query Builder, but just like other SQL databases enabling SQL access would imply access to any dataset in that DB

Switching BigQuery to work this way could be a bit tricky however

**Comment by VikramTiwari** [comment]
I agree that parsing SQL is fragile security. But it's a possible solution in the current case.

If I understand correctly, treating each BigQuery database as a Metabase database will just push this to another level. Then Metabase might need to build dataset level permissions, the same way there are table level permissions for other databases. Effectively the problem still remains there if I want to restrict access to a dataset.

I understand that it won't confuse the user since it will be clear that access is there because it's a database. But the issue will still be there? @camsaul Am I missing something?

**Comment by camsaul** [comment]
If we treated datasets as though they were schemas, then permissions is already ready to go and we don't need to build anything. That's one of the reasons I think such a change might make sense 😉

**Comment by VikramTiwari** [comment]
I am having a hard time visualizing it, but I will take your word for it. 👍  Do you think we should treat this issues as a security/priority issue or no?

**Comment by salsakran** [comment]
Does BQ let you control a user's access to specific datasets?

Seems like the more "correct" solution here is to enforce these constraints (eg, a "DB" being a single unit of data that can't be "brokenout of"), rather than try to paper over it on our end.

**Comment by salsakran** [comment]
More specifically, each "DB" in metabase should have a different user + ACL for that user that locks things down to just the dataset in question. 

Even if we do go down the schema route (which I think sounds good), I think it's better practice to lock things down as far upstream as possible.

**Comment by VikramTiwari** [comment]
To answer the question:
- Yes, but this is a setting while one is creating credentials for the BigQuery to connect to. If those credentials have the proper scope then they should restrict access. Ref: https://cloud.google.com/bigquery/docs/access-control#predefined_roles_details

**Comment by VikramTiwari** [comment]
Just an observation:

Shouldn't Metabase be focusing a bit more on enhancing BigQuery and GA support? The reason I say this is cause both are essentially free databases (BigQuery now offers 10G free storage). The more feature complete Metabase is for GA and BQ, the easier adoption should be.

**Comment by camsaul** [comment]
@VikramTiwari Lots of databases we support are free. (Postgres, Mongo, MySQL, etc.)

We are committed to continuing to improve our BigQuery support; our last major release (0.24) fixed over 8 BigQuery-specific issue. But we're a small team, so unfortunately we can't fix every issue the day after it's opened.

**Comment by VikramTiwari** [comment]
I meant free as in hosted for free. And this statement wasn't supposed to be for this issue only. There are a lot of issues which I see open for GA and BQ. Maybe it's just me since I majorly use them as data sources.

I would love to contribute with code, but I have no experience in Clojure. I would be looking into it though.

**Comment by Saurabh-Shrivastava-Narvar** [comment]
Any progress on this? It's been over a year - we are facing same issue. Not able to restrict permissions at dataset level is stopping us rolling out metabase for bigquery.

**Comment by m-lce** [comment]
> When we first added the BigQuery driver there was some discussion about whether we should treat each dataset as a different database or treat your entire project as a single database (and each dataset as a "schema" within that database)
> 
> there were a lot of good arguments for both approaches, but perhaps the latter would have made more sense since it more closely models reality. This is something we might want to consider changing to avoid confusing cases like this
I think that would be great to have the latter because that’s a quite common use case to join tables of different datasets and I found it as big limitation of Metabase usage with Big Query

**Comment by flamber** [comment]
Related to #6338 and #10525.
As for the database vs schema approach, there's a separate issue for that #7250. Upvote by clicking :+1: on the first post

**Comment by ribrdb** [comment]
As far as parsing sql securely, I believe google's zeta sql parser is supposed to support all bigquery syntax: https://github.com/google/zetasql
